cpp/src/reductions/scan/scan.cuh

```diff
@@ -28,38 +28,52 @@ namespace cudf {
 namespace detail {
 
 // logical-and scan of the null mask of the input view
-rmm::device_buffer mask_scan(const column_view& input_view,
-                             cudf::scan_type inclusive,
+rmm::device_buffer mask_scan(column_view const& input_view,
+                             scan_type inclusive,
                              rmm::cuda_stream_view stream,
                              rmm::mr::device_memory_resource* mr);
 
+std::unique_ptr<column> inclusive_rank_scan(column_view const& order_by,
+                                            rmm::cuda_stream_view stream,
+                                            rmm::mr::device_memory_resource* mr);
+
+std::unique_ptr<column> inclusive_dense_rank_scan(column_view const& order_by,
+                                                  rmm::cuda_stream_view stream,
+                                                  rmm::mr::device_memory_resource* mr);
+
 template <template <typename> typename DispatchFn>
 std::unique_ptr<column> scan_agg_dispatch(const column_view& input,
                                           std::unique_ptr<aggregation> const& agg,
                                           null_policy null_handling,
                                           rmm::cuda_stream_view stream,
                                           rmm::mr::device_memory_resource* mr)
 {
-  CUDF_EXPECTS(
-    is_numeric(input.type()) || is_compound(input.type()) || is_fixed_point(input.type()),
-    "Unexpected non-numeric or non-string type.");
+  if (agg->kind == aggregation::RANK || agg->kind == aggregation::DENSE_RANK) {
+    CUDF_EXPECTS(input.type().id() != type_id::LIST, "Unsupported list type in rank aggregation.");
+  } else {
+    CUDF_EXPECTS(
+      is_numeric(input.type()) || is_compound(input.type()) || is_fixed_point(input.type()),
+      "Unexpected non-numeric or non-string type.");
+  }
```

Comment:
This error checking shouldn't be done manually like this. It should be done after type dispatching.

Reply:
The reason I added these was because the dispatch was erring silently and generating garbage result. For example, I could run the scan with a nested list column with the element comparator returning false whenever the types were supported. What would be the best way to make the element fail loudly to avoid this case?

Reply:
Typically the dispatch functors should handle such error checks. If they don't then I think we should check it here. If the existing ones already did that while your new functors don't, move your error check to your functors.

Reply:
Mixed up the type dispatch you were referencing, I've moved the type checking down inside the different rank scan functions.